internal/provider/helpers.go

```diff
@@ -0,0 +1,50 @@
+package provider
+
+import (
+	"fmt"
+
+	dataschema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	resourceschema "github.com/hashicorp/terraform-plugin-framework/resource/schema"
+)
+
+// convertSchemaAttributes helps to convert resource schema to datasource schema.
+// All attributes marked as Computed, not Required and not Optional.
+// You can modify any of the attributes later.
+func convertSchemaAttributes(attrs map[string]resourceschema.Attribute, diagn diag.Diagnostics) map[string]dataschema.Attribute {
+	res := make(map[string]dataschema.Attribute)
+
+	for name, attrInterface := range attrs {
+		switch attr := attrInterface.(type) {
+		case resourceschema.StringAttribute:
+			res[name] = convertStringAttribute(attr)
+		case resourceschema.SingleNestedAttribute:
+			res[name] = convertSingleNestedAttribute(attr, diagn)
+		default:
+			diagn.AddError("can not convert resource attribute to datasource attribute", fmt.Sprintf("unsupported type for attribute %q: %v", name, attr))
```

Comment:
That's a unusual idiomatic way to return error.
I would suggest one of two approaches:
 * Make this function more `clean` and return a usual pair `(map[string]dataschema.Attribute, error)`
 * Make this function a little bit `clean` with return tuple like `(map[string]dataschema.Attribute, diag.Diagnostics)`

Reply:
I agree with you. But it looks like that's almost common way in terraform world.
Such behaviour brings one useful benefit. You can write code like
```go
s := someStructure{
    arg1: "val1",
    arg2: func(...),
}
```
instead of
```go
val2, err := func(...)
if err != nil ....

s := someStructure{
    arg1: "val1",
    arg2: val2,
}
```

Reply:
Well, we are in the golang land.
I thought we all adore `if err != nil` statement and clean functions, isn't it?

I agree that the second option is heavier but also it clearer.
Also, idea of immediate if clauses is to have opportunity to call `return nil, diags` in cases when next block code doesn't make sense because of error.

Let's stick to one style (2nd).